GAWB-3314: create pet SA in google if it only exists in OpenDJ (also add missing key resiliency)

MatthewBemis opened the pull request:
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

MatthewBemis committed:
create pet SA in google if it only exists in OpenDJ

MatthewBemis commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:
```diff
@@ -228,7 +229,7 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
       result <- maybePet match {
         case Some(pet) => googleKeyCache.removeKey(pet, keyId)
-        case none => Future.successful(())
+        case None => Future.successful(())
```
fortunately this worked before...

jmthibault79 replied:
ha

MatthewBemis commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:
```diff
@@ -188,21 +188,22 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
-        petSA.flatMap { serviceAccount =>
-          // Set up the service account with the necessary permissions
-          val petServiceAccount = PetServiceAccount(PetServiceAccountId(user.id, project), serviceAccount)
+    // First find or create the service account in Google, which generates a unique id and email
```
gpalloc and other "rogue" beings may delete a pet SA in google, leaving a reference to it behind in Sam.

previously we would check to see if the pet SA existed in OpenDJ, and assume everything was cool in google if it existed. since that's not always going to be the case, i've switched up the ordering to first getOrCreate the pet in google and then create the pet in opendj if needed.

per my local logs, the extra google call appears to add about ~500ms to each call to get an existing key.

jmthibault79 replied:
Thanks!  Add a comment indicating why we may need to recreate the GooglePet even if it's in OpenDJ, so we don't accidentally "optimize" this in the future

jmthibault79 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:
```diff
@@ -188,21 +188,22 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
+    val petSA = googleIamDAO.getOrCreateServiceAccount(project, petSaID, petSaDisplayName)
+    petSA.flatMap { serviceAccount =>
+      // Set up the service account with the necessary permissions
+      val petServiceAccount = PetServiceAccount(PetServiceAccountId(user.id, project), serviceAccount)
```
This is thrown away if it's in OpenDJ so we should move that to the `None` below

jmthibault79 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala:
```diff
@@ -188,21 +188,22 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
   def createUserPetServiceAccount(user: WorkbenchUser, project: GoogleProject): Future[PetServiceAccount] = {
     val (petSaID, petSaDisplayName) = toPetSAFromUser(user)
```
I was wondering what the difference was between`petSaID` and the `PetServiceAccountId` below.  Turns out this is a *name* rather than an ID.  Please rename it as such.

jmthibault79 commented:
Looks good.  I would also like to see an integration test with a rogue pet deletion and confirmation that `getPetServiceAccountKey()` still works afterward.

MatthewBemis committed:
initial round of PR comments

MatthewBemis committed:
add integration test

MatthewBemis commented:
@jmthibault79 i've added an integration test but i'm concerned now that it could have side effects on other tests on other fiabs since it is using the default project. this would likely be rare, but it's still a possibility.

i've updated it locally to use GPAlloc but there is a permissions issue when deleting the pet SA from google. i'll need to dive in further to see what's wrong. depending on the amount of work required to get it to work with GPAlloc, we may want to merge this with the test ignored and a follow-up to re-enable it with GPAlloc.

MatthewBemis committed:
some cleanup, genericize getFieldFromJson

MatthewBemis committed:
clean up some imports

jmthibault79 commented:
Thanks!  I wonder if `googleIamDAO` is trying to remove it as the test-runner user and they don't have the right perms

MatthewBemis committed:
add resiliency in case the key doesn't actually exist in google anymore

MatthewBemis commented:
@jmthibault79 in the latest commit i also added some resiliency for the case we originally talked about where a key has been deleted directly from the SA but still exists in the Google bucket. this wasn't related to the errors you were seeing, but i did end up noticing a case where things could go wrong

MatthewBemis committed:
rename ambiguous variable

MatthewBemis commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleKeyCache.scala:
```diff
@@ -73,11 +73,17 @@ class GoogleKeyCache(val googleIamDAO: GoogleIamDAO, val googleStorageDAO: Googl
+    //The key may exist in the Google bucket cache, but could have been deleted from the SA directly
+    val keyExistsForSA = serviceAccountKeys.exists(_.id.value.contentEquals(keyId))
+
+    if(keyRetired || !keyExistsForSA) furnishNewKey(pet)
```
also we do not need to worry about deleting the key, we can wait for the lifecycle to pick them up

dvoet replied:
what happens if key is already gone from google when deleted from bucket?

MatthewBemis replied:
it's handled without an issue:

1) pub/sub message is triggered
2) sam tries to delete it, gets a 404 which recovers gracefully in HttpGoogleIamDAO 
3) sam sees this as a success and acks the message

dvoet reviewed (approved):
make sure error is handled when deleting a key from google and it does not exist

MatthewBemis committed:
clean up the keys in tests just in case

jmthibault79 reviewed (approved):
ok to ignore the test for now, to make other tests more reliable

MatthewBemis committed:
ignore integration test and add documentation about why
